packages/@aws-cdk/aws-apigateway/lib/authorizers/authorizer-base.ts

```diff
@@ -0,0 +1,91 @@
+import { Construct, Lazy, Resource, ResourceProps } from '@aws-cdk/core';
+import { IAuthorizer } from '../authorizer';
+import { AuthorizationType, Method } from '../method';
+import { IRestApi } from '../restapi';
+
+const AUTHORIZER_BASE_SYMBOL = Symbol.for('@aws-cdk/apigateway.AuthorizerBase');
+
+/**
+ * Base class for all custom authorizers
+ */
+export abstract class AuthorizerBase extends Resource implements IAuthorizer {
+
+  /**
+   * Check if the given object is of type CustomAuthorizer
+   * @internal
+   */
+  public static _isAuthorizer(x: any): x is AuthorizerBase {
+    return x !== null && typeof(x) === 'object' && AUTHORIZER_BASE_SYMBOL in x;
+  }
+
+  /**
+   * The authorizer ID.
+   * @attribute
+   */
+  public abstract readonly authorizerId: string;
+
+  private _restApi?: IRestApi;
+
+  constructor(scope: Construct, id: string, props?: ResourceProps) {
+    super(scope, id, props);
+
+    Object.defineProperty(this, AUTHORIZER_BASE_SYMBOL, { value: true });
+  }
+
+  /**
+   * Called when this authorizer needs to be bound to a specific {@link Method}.
+   *
+   * @returns configuration that may be needed by the caller to record this bind.
+   * @throws when this authorizer is already bound to a Method from a different RestApi instance.
+   *
+   * @internal
+   */
+  public _bind(method: Method) {
+    const restApi = method.restApi;
+    if (this._restApi && this._restApi !== restApi) {
+      throw new Error(`Cannot attach authorizer [${this.node.uniqueId}] to RestApi [${restApi.node.uniqueId}]. `
+        + `Already attached to RestApi [${this._restApi.node.uniqueId}]`);
+    }
+    this._restApi = restApi;
+    return this.authorizerConfig(method);
+  }
+
+  /**
+   * Returns a lazy that will resolve to the restApiId at the time of synthesis.
+   */
+  protected get restApiId(): string {
+    return Lazy.stringValue({
+      produce: () => {
+        if (this._restApi) {
+          return this._restApi.restApiId;
+        } else {
+          throw new Error(`LambdaAuthorizer [${this.physicalName}] is not associated with a RestApi construct`);
+        }
+      }
+    });
+  }
+
+  /**
+   * Configuration returned back to the {@link Method} that it may use to record the bind.
+   *
+   * @param method the method to which this authorizer is being bound.
+   * @returns configuration that may be needed by the caller to record this bind.
+   */
+  protected abstract authorizerConfig(method: Method): AuthorizerConfig;
```

Comment:
Why not just let users override `bind` and call super? (that's a common pattern)

Reply:
During binding we need to get the `RestApi` from the method and associate it with the `Authorizer`. `bind` needs to be delegated as you suggest but also some work needs to be performed at the base class as well. What's a good pattern for this?

Reply:
The derived class can just call the base class ‘super.bind()’, no?

Reply:
Sure I guess yes. But it **must not** forget.

I'll switch to change to that.

Reply:
Ahh, more modeling problems. 

First, base class **must** implement bind() which means it should (doesn't have to) be abstract. Second, if the base class has an implementation, it needs to return an instance of `AuthorizerConfig` which it can't since there are no 'correct' values to return. I could make all the members optional and return an empty which isn't great either.